### PR TITLE
Created wrapper + tests for the Kademlia Library

### PR DESCRIPTION
**Pull request by thompalex**
Modified the peer class by creating a callback function for rpc_store in the Kademlia library to access the key and value.

NewServer overrides the _create_protocol method from Kademlia's network.Server to create an instance of our protocol.

Added tests for our NewServer and NewProtocol -> 100% testing coverage of this code.

### CODE REVIEW

File: src/main/py/peer/peer.py [hunk 1]
```diff
@@ -80,16 +82,40 @@ def commission_art_piece(self) -> None:
...
+        """
+        Callback function for when data is stored.
+        Args:
+            key (bytes): The key to store.
+            value (bytes): The value to store.
+        """
+        self.logger.info("Data stored with key: %s", key)
+        self.logger.info("Data stored with value: %s", value)
+        artwork_object = pickle.loads(value)
+        if isinstance(artwork_object, Artwork):
```
Comment (zimmermatt) [inline src/main/py/peer/peer.py:105]: Might be useful to have an `else` that logs that a non-artwork instance, if that's unexpected.

File: src/main/py/server/network.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python3
+""" This module contains the NewServer class. """
+
+from kademlia.network import Server
+from server.protocol import NewProtocol
+
+
+class NewServer(Server):
```
Comment (zimmermatt) [inline src/main/py/server/network.py:8]: Let's brainstorm some more descriptive names :)

File: src/main/py/server/protocol.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+"""
+Module for Kademlia library rpc_store() callback
+"""
+from kademlia.protocol import KademliaProtocol
+
+
+class NewProtocol(KademliaProtocol):
```
Comment (zimmermatt) [inline src/main/py/server/protocol.py:7]: Let's brainstorm some more descriptive names :)

File: src/main/py/server/protocol.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        """
+        Initialize a new instance of NewProtocol.
+
+        Args:
+            source_node (Node): The source node.
+            storage (Storage): The storage object.
+            ksize (int): The k parameter for Kademlia.
+            callback_function (function): The callback function to be called on rpc_store
+        """
+        self.callback_function = callback_function
```
Comment (zimmermatt) [inline src/main/py/server/protocol.py:22]: The name for the callback in `NewServer` is better IMO.

File: src/test/py/server/new_server_test.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    def test_init(self):
+        """Test the __init__ method of NewServer"""
+        self.assertEqual(self.node1.store_callback, self.mock_store_callback)
+
+    async def test_store(self):
+        """Test the store method of NewServer"""
+        await self.node1.listen(8468)
+        await self.node2.listen(8469)
+        await self.node2.bootstrap([("127.0.0.1", 8468)])
```
Comment (zimmermatt) [inline src/test/py/server/new_server_test.py:33]: Let's see how it goes, but spinning up/tearing down servers may be a source of instability in the test suite.

File: src/test/py/server/new_server_test.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        self.node2 = NewServer(self.mock_store_callback)
+
+    def test_init(self):
+        """Test the __init__ method of NewServer"""
+        self.assertEqual(self.node1.store_callback, self.mock_store_callback)
+
+    async def test_store(self):
+        """Test the store method of NewServer"""
+        await self.node1.listen(8468)
+        await self.node2.listen(8469)
```
Comment (zimmermatt) [inline src/test/py/server/new_server_test.py:32]: Please use ports in the dynamic range: 49152 - 65535.

See: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt

**Review by zimmermatt — CHANGES_REQUESTED** [review]

File: src/main/py/peer/peer.py [hunk 7]
```diff
@@ -47,35 +47,57 @@ async def send_deadline_reached(self, commission: Artwork) -> None:
...
+                commission.get_key(), pickle.dumps(commission)
+            )
+            if set_success:
+                self.logger.info("Commission sent")
+                self.commissions.append(commission)
+            else:
+                self.logger.error("Commission failed to send")
+        except TypeError:
+            self.logger.error("Commission type is not pickleable")
         await self.setup_deadline_timer(commission)
```
Comment (zimmermatt) [inline src/main/py/peer/peer.py:94]: This looks like it belongs in the `if set_success:` block as well.

**Review by zimmermatt — CHANGES_REQUESTED** [review]
Looking good! Just one comment to take a look at.

Also, it will be helpful if you can point me at the commit with the text you'd like for the squashed merge that covers the entirety of the PR. Or, please write one up that I can use when you push this final change.

**Comment by thompalex** [comment]
I think the following will make for a good commit message for the squash:

Subject: Created NotifyingServer and NotificationProtocol for our interaction with the Kademlia module

- NotificationServer and NotificationProtocol run the Kademlia module in the same way as usual, but they take in a callback function for the rpc_store method so that we can listen for and react to new items stored in the DHT. This will allow us to use the DHT as a sort of messaging system between peers.
- Adjusted the current peer class to await necessary asynchronous functions and modified our tests as needed to account for this change.
- Created a preliminary callback function to call on new store requests in our peer class. This callback will evolve to handle different types of "messages" or objects in the DHT as we progress with our goals.

File: src/test/py/peer/peer_test.py [hunk 8]
```diff
@@ -86,21 +87,28 @@ async def test_commission_art_piece(self, mock_input):
...
+            return_value=MagicMock(
+                call_later=lambda *args: setattr(
+                    self, "deadline_task", asyncio.create_task(args[2])
+                )
+            ),
+        ):
+            self.test_logger.debug(mock_input)
+            self.peer.node = self.mock_node
+            await self.peer.commission_art_piece()
+            self.assertEqual(len(self.peer.commissions), 1)
```
Comment (zimmermatt) [inline src/test/py/peer/peer_test.py:101]: At some point later, a test should be added that tests the failure case and asserts that the commissions count does not increase.

**Comment by zimmermatt** [comment]
> I think the following will make for a good commit message for the squash:
> 
> Subject: Created NotifyingServer and NotificationProtocol for our interaction with the Kademlia module
> 
> * NotificationServer and NotificationProtocol run the Kademlia module in the same way as usual, but they take in a callback function for the rpc_store method so that we can listen for and react to new items stored in the DHT. This will allow us to use the DHT as a sort of messaging system between peers.
> * Adjusted the current peer class to await necessary asynchronous functions and modified our tests as needed to account for this change.
> * Created a preliminary callback function to call on new store requests in our peer class. This callback will evolve to handle different types of "messages" or objects in the DHT as we progress with our goals.

Looks good. I'm going to make a couple of minor revisions when I merge to comply with the style guide while maintaining the core of your version. Mainly:
* subject <= 50 chars
* imperative mood (guide says "in the subject line" but it also applies to body)

**Review by zimmermatt — APPROVED** [review]

**Pull request merged**